src/security/PeerConnector.cc

```diff
@@ -694,39 +786,85 @@ Security::PeerConnector::certDownloadingDone(SBuf &obj, int downloadStatus)
         return;
     }
 
-    srvBio->holdRead(false);
-    negotiateSsl();
+    resumeNegotiation();
 }
 
+void
+Security::PeerConnector::handleMissingCertificates(const TlsNegotiationDetails &ed)
+{
+    auto &sconn = *fd_table[serverConnection()->fd].ssl;
+
+    // We download the missing certificate(s) once. We would prefer to clear
+    // this right after the first validation, but that ideal place is _inside_
+    // OpenSSL if validation is triggered by SSL_connect(). That function and
+    // our OpenSSL verify_callback function (\ref OpenSSL_vcb_disambiguation)
+    // may be called multiple times, so we cannot reset there.
+    auto &callerHandlesMissingCertificates = Ssl::VerifyCallbackParameters::At(sconn).callerHandlesMissingCertificates;
+    Must(callerHandlesMissingCertificates);
+    callerHandlesMissingCertificates = false;
+
+    if (!computeMissingCertificateUrls(sconn))
+        return handleNegotiateError(ed);
+
+    suspendNegotiation(ed);
+
+    assert(!urlsOfMissingCerts.empty());
+    startCertDownloading(urlsOfMissingCerts.front());
+    urlsOfMissingCerts.pop();
+}
+
+/// calculates URLs of the missing intermediate certificates or returns false
 bool
-Security::PeerConnector::checkForMissingCertificates()
+Security::PeerConnector::computeMissingCertificateUrls(const Connection &sconn)
 {
-    // Check for nested SSL certificates downloads. For example when the
-    // certificate located in an SSL site which requires to download a
-    // a missing certificate (... from an SSL site which requires to ...).
+    const auto certs = SSL_get_peer_cert_chain(&sconn);
+    if (!certs) {
+        debugs(83, 3, "nothing to bootstrap the fetch with");
+        return false;
+    }
+    debugs(83, 5, "server certificates: " << sk_X509_num(certs));
 
-    const Downloader *csd = (request ? request->downloader.get() : nullptr);
-    if (csd && csd->nestedLevel() >= MaxNestedDownloads)
+    const auto ctx = getTlsContext();
+    if (!Ssl::missingChainCertificatesUrls(urlsOfMissingCerts, certs, ctx) || urlsOfMissingCerts.empty()) {
```

Comment:
Something went wrong here. We should not expect the caller to remember to check urlsOfMissingCerts.empty() after a successful missingChainCertificatesUrls() return. I have not checked all the details, but I suspect we should move that second check inside missingChainCertificatesUrls() _and_ provide debugging for `false` cases there. This code will then become:

```suggestion
    if (!Ssl::missingChainCertificatesUrls(urlsOfMissingCerts, certs, ctx))
        return false; // missingChainCertificatesUrls() reports the exact reason
    assert(!urlsOfMissingCerts.empty());
```

Reply:
commit b96081c35